handler/actions/src/main/java/io/knotx/fragments/handler/action/http/response/ResponseProcessor.java

```diff
@@ -0,0 +1,167 @@
+    ActionRequest actionRequest = createActionRequest(endpointRequest);
+    ActionResponse actionResponse = createErrorActionResponse(endpointResponse);
+    return ActionPayload.noResult(actionRequest, actionResponse);
+  }
+
+  private ActionRequest createActionRequest(EndpointRequest endpointRequest) {
+    ActionRequest request = new ActionRequest(HTTP_ACTION_TYPE, endpointRequest.getPath());
+    request.appendMetadata(METADATA_HEADERS_KEY,
+        MultiMapTransformer.toJson(endpointRequest.getHeaders()));
+    return request;
+  }
+
+  private ActionResponse createSuccessActionResponse(EndpointResponse endpointResponse) {
+    return ActionResponse.success()
+        .appendMetadata(METADATA_STATUS_CODE_KEY,
+            String.valueOf(endpointResponse.getStatusCode().code()))
+        .appendMetadata(METADATA_HEADERS_KEY,
+            MultiMapTransformer.toJson(endpointResponse.getHeaders()));
+  }
+
+  private ActionResponse createErrorActionResponse(EndpointResponse endpointResponse) {
+    return ActionResponse
+        .error(endpointResponse.getStatusCode().toString(), endpointResponse.getStatusMessage())
+        .appendMetadata(METADATA_STATUS_CODE_KEY,
+            String.valueOf(endpointResponse.getStatusCode().code()))
+        .appendMetadata(METADATA_HEADERS_KEY,
+            MultiMapTransformer.toJson(endpointResponse.getHeaders()));
+  }
+
+  private Object tryToRetrieveResultFrom(EndpointResponse endpointResponse,
+      HttpActionLogger httpActionLogger) {
+    try {
+      return retrieveResultFrom(endpointResponse);
+    } catch (Exception exception) {
+      httpActionLogger.onResponseProcessingFailed(exception);
+      throw exception;
+    }
+  }
+
+  private Object retrieveResultFrom(EndpointResponse response) {
+    if (isForceJson || isJsonPredicate || isContentTypeHeaderJson(response)) {
+      return bodyToJson(response.getBody().toString());
+    } else {
+      return response.getBody().toString();
+    }
+  }
+
+  private boolean isContentTypeHeaderJson(EndpointResponse endpointResponse) {
+    String contentType = endpointResponse.getHeaders().get(CONTENT_TYPE);
+    return contentType != null && contentType.contains(APPLICATION_JSON);
+  }
+
+  private Object bodyToJson(String responseBody) {
+    if (StringUtils.isBlank(responseBody)) {
+      return new JsonObject();
+    } else if (responseBody.startsWith("[")) {
+      return new JsonArray(responseBody);
+    } else {
+      return new JsonObject(responseBody);
+    }
+  }
+
+  private String getErrorTransition(EndpointResponse endpointResponse) {
+    if (isTimeout(endpointResponse)) {
+      return TIMEOUT_TRANSITION;
+    } else {
+      return ERROR_TRANSITION;
+    }
+  }
+
+  private boolean isSuccess(EndpointResponse response) {
+    return HttpStatusClass.SUCCESS.contains(response.getStatusCode().code());
+  }
+
+  private boolean isTimeout(EndpointResponse response) {
+    return HttpResponseStatus.REQUEST_TIMEOUT == response.getStatusCode();
```

Comment:
It is a client error status code: https://tech-banker.com/how-to-fix-a-408-request-timeout-error/. What about `GATEWAY_TIMEOUT`? Should we support this too?

Reply:
I think this is a separate case. Gateway timeout indicates that the response was received, but timeout occured in the called endpoint's internals. The semantics of these seem to me as different since the `REQUEST_TIMEOUT` might be caused by network/hardware issues on Knot.x side, whereas `GATEWAY_TIMEOUT` indicates there is a problem with the invoked endpoint.

Reply:
OK, I think that server timeouts should be also validated. Issue #103.